app/src/main/java/org/mozilla/fenix/TelemetryMiddleware.kt

```diff
@@ -0,0 +1,106 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix
+
+import androidx.annotation.VisibleForTesting
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.ContentAction
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.selector.normalTabs
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.base.log.logger.Logger
+import org.mozilla.fenix.components.metrics.Event
+import org.mozilla.fenix.components.metrics.MetricController
+import org.mozilla.fenix.search.telemetry.ads.AdsTelemetry
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * [Middleware] to record telemetry in response to [BrowserAction]s.
+ *
+ * @property settings reference to the application [Settings].
+ * @property adsTelemetry reference to [AdsTelemetry] use to record search telemetry.
+ * @property metrics reference to the configured [MetricController] to record general page load events.
+ */
+class TelemetryMiddleware(
+    private val settings: Settings,
+    private val adsTelemetry: AdsTelemetry,
+    private val metrics: MetricController
+) : Middleware<BrowserState, BrowserAction> {
+
+    private val logger = Logger("TelemetryMiddleware")
+
+    @VisibleForTesting
+    internal val redirectChains = mutableMapOf<String, RedirectChain>()
+
+    /**
+     * Utility to collect URLs / load requests in between location changes.
+     */
+    internal class RedirectChain(internal val root: String) {
+        internal val chain = mutableListOf<String>()
+
+        fun add(url: String) {
+            chain.add(url)
+        }
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
```

Comment:
So our `action` will always be a `ContentAction`, and `next(action)` will always give us a `TabListAction`? Can we use the more specific types?

Reply:
Ah no the action is always a `BrowserAction` here. The type of action is specified by the store. For example the type signature of the `BrowserStore` (which this middleware is for) is `Store<BrowserState, BrowerAction>`, basically defining that this store manages `BrowserState` which gets updated by `BrowserAction`s. 

`ContentAction`, `TabListAction` etc. are all subtypes of `BrowserAction` so in order to intercept actions on the store we need the top-level `BrowserAction`.

See: https://github.com/mozilla-mobile/android-components/blob/master/components/browser/state/src/main/java/mozilla/components/browser/state/store/BrowserStore.kt#L25

This is the same pattern for all stores in Fenix, say the `LoginsFragmentStore` which is a `Store<LoginsListState, LoginsAction>`. If we wanted to intercept actions there we'd use a `Middleware<LoginsListState, LoginsAction>`.

See: https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/settings/logins/LoginsFragmentStore.kt#L45

Reply:
> and next(action) will always give us a TabListAction?

Using a middleware it's possible to intercept actions before and after these actions are processed by store (and the corresponding state is updated). `next(action)` is what propagates actions to the next middleware in line and then ultimately to the store. 

So the actions we process before the call to `next(action)` haven't yet reached the store, which is useful if we want to read the state or apply side effects before the state has changed. So, this is useful for preprocessing i.e. "before this action is applied do this." 

The actions we process after the call to `next(action)` have already reached the store and the state has already been changed. So, this is useful for postprocessing i.e. "after this action has been applied do this."

Reply:
We modelled this after https://redux.js.org/advanced/middleware which has a good introduction as well, but happy to chat!